### PR TITLE
Add support for password resets

### PR DESCRIPTION
**Pull request by petschekr**
* Adds support for password resets (**please look over implementation and try to break**)
* HTMLifies emails for verification and resets for easier linking and interpolation support
* Fixes submit error when sign up is clicked more than once
* Checks if submissions are open and allowed in the application and confirmation submission API

### CODE REVIEW

File: client/login.html [hunk 1]
```diff
@@ -9,6 +9,18 @@
 		<link rel="stylesheet" href="/css/login.css" />
 		<link rel="stylesheet" href="/css/theme.css" />
 		<script src="https://use.fontawesome.com/efaca8fe24.js"></script>
+		<script>
+			document.addEventListener("DOMContentLoaded", () => {
+				const signUpButton = document.querySelectorAll('input[type="submit"]')[0];
+				const logInButton = document.querySelectorAll('input[type="submit"]')[1];
+				document.querySelector('form[action="/auth/signup"').addEventListener("submit", () => {
```
Comment (petschekr) [inline client/login.html]: Nope! Thanks for catching that

File: client/login.html [hunk 2]
```diff
@@ -9,6 +9,18 @@
 		<link rel="stylesheet" href="/css/login.css" />
 		<link rel="stylesheet" href="/css/theme.css" />
 		<script src="https://use.fontawesome.com/efaca8fe24.js"></script>
+		<script>
+			document.addEventListener("DOMContentLoaded", () => {
+				const signUpButton = document.querySelectorAll('input[type="submit"]')[0];
+				const logInButton = document.querySelectorAll('input[type="submit"]')[1];
+				document.querySelector('form[action="/auth/signup"]').addEventListener("submit", () => {
+					signUpButton.disabled = true;
```
Comment (illegalprime) [inline client/login.html]: Why not disable both (`signUpButton` and `logInButton`) so you can't double click it?

File: server/routes/auth.ts [hunk 3]
```diff
@@ -237,11 +237,11 @@ passport.use(new LocalStrategy({
 			return;
 		}
 		// Send verification email (hostname validated by previous middleware)
-		let link = `http${request.secure ? "s" : ""}://${request.hostname}:${getExternalPort(request)}/auth/verify/${user.localData!.verificationCode}`;
-		let text =
-`Hi ${user.name},
+		let link = createLink(request, `/auth/verify/${user.localData!.verificationCode}`);
+		let markdown =
+`Hi {{name}},
```
Comment (illegalprime) [inline server/routes/auth.ts]: Is this supposed to be `{{var}}` instead of `${var}`?

File: server/routes/auth.ts [hunk 4]
```diff
@@ -396,6 +412,117 @@ authRoutes.get("/verify/:code", async (request, response) => {
...
+	}
+
+	user.localData.resetRequested = true;
+	user.localData.resetRequestedTime = new Date();
+	user.localData.resetCode = crypto.randomBytes(32).toString("hex");
+
+	// Send reset email (hostname validated by previous middleware)
+	let link = createLink(request, `/auth/forgot/${user.localData.resetCode}`);
+	let markdown =
+`Hi {{name}},
```
Comment (illegalprime) [inline server/routes/auth.ts]: `name` isn't in scope

File: server/routes/auth.ts [hunk 5]
```diff
@@ -396,6 +412,117 @@ authRoutes.get("/verify/:code", async (request, response) => {
...
+	}
+});
+authRoutes.post("/forgot/:code", validateAndCacheHostName, postParser, async (request, response) => {
+	let user = await User.findOne({ "localData.resetCode": request.params.code });
+	if (!user) {
+		request.flash("error", "Invalid password reset code");
+		response.redirect("/login");
+		return;
+	}
+	else if (!user.localData!.resetRequested || Date.now() - user.localData!.resetRequestedTime.valueOf() > 1000 * 60 * 60) {
```
Comment (illegalprime) [inline server/routes/auth.ts]: make expiration time a constant and part of the config file

File: server/routes/auth.ts [hunk 6]
```diff
@@ -396,6 +412,117 @@ authRoutes.get("/verify/:code", async (request, response) => {
...
+	let password1: string | undefined = request.body.password1;
+	let password2: string | undefined = request.body.password2;
+	if (!password1 || !password2) {
+		request.flash("error", "Missing new password or confirm password");
+		response.redirect(`/auth${request.url}`);
+		return;
+	}
+	if (password1 !== password2) {
+		request.flash("error", "Passwords must match");
+		response.redirect(`/auth${request.url}`);
```
Comment (illegalprime) [inline server/routes/auth.ts]: use `path.join`

File: server/routes/auth.ts [hunk 7]
```diff
@@ -396,6 +412,117 @@ authRoutes.get("/verify/:code", async (request, response) => {
...
+		await user.save();
+		response.redirect("/login");
+		return;
+	}
+
+	let password1: string | undefined = request.body.password1;
+	let password2: string | undefined = request.body.password2;
+	if (!password1 || !password2) {
+		request.flash("error", "Missing new password or confirm password");
+		response.redirect(`/auth${request.url}`);
```
Comment (illegalprime) [inline server/routes/auth.ts]: `path.join`

**Comment by illegalprime** [comment]
Let's remember that merging this to `master` **will** deploy to everyone, so we gotta be careful nowadays.

File: server/routes/auth.ts (hunk 3 above)
Comment (petschekr) [inline server/routes/auth.ts]: This uses the built-in email variable interpolation support which is why the `{{}}` syntax is used

File: server/routes/auth.ts (hunk 4 above)
Comment (petschekr) [inline server/routes/auth.ts]: Also uses the email variable interpolation from `common.ts`

File: server/routes/auth.ts (hunk 5 above)
Comment (petschekr) [inline server/routes/auth.ts]: Good idea

File: server/routes/auth.ts (hunk 7 above)
Comment (petschekr) [inline server/routes/auth.ts]: :+1:

File: client/login.html (hunk 2 above)
Comment (petschekr) [inline client/login.html]: I don't exactly follow...

Comment (illegalprime) [inline client/login.html]: I mean disable all buttons so you can't quickly double click any button once you click one button (so you don't send multiple requests on accident)

**Comment by petschekr** [comment]
@illegalprime Updated

**Pull request merged**